src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnRunner.java

```diff
@@ -291,15 +300,26 @@ private SpawnResult downloadAndFinalizeSpawnResult(
       SpawnExecutionContext context,
       RemoteOutputsMode remoteOutputsMode)
       throws ExecException, IOException, InterruptedException {
+    // If one output of a spawn is a top level output then download all outputs. Spawns
+    // are typically structured in a way that either all or no outputs are top level and
+    // it's much simpler to implement under this assumption.
+    boolean spawnHasTopLevelOutputs = remoteOutputsMode.downloadToplevelOutputsOnly() &&
+        hasTopLevelOutputs(spawn.getOutputFiles(), topLevelOutputs);
     SpawnResult.Status actionStatus =
         actionResult.getExitCode() == 0 ? Status.SUCCESS : Status.NON_ZERO_EXIT;
+    RemoteOutputsMode effectiveOutputsMode;
```

Comment:
The lines 310-319 could be nicely extracted to a separate method: defineOutputsMode or smth like this.